src/storage/txn/flow_controller.rs

```diff
@@ -0,0 +1,1055 @@
+        })
+        .unwrap();
+
+        Self {
+            discard_ratio,
+            limiter,
+            enabled: Arc::new(AtomicBool::new(config.enable)),
+            tx,
+            handle: Some(checker.start(rx, flow_info_receiver)),
+        }
+    }
+
+    pub fn should_drop(&self) -> bool {
+        let ratio = self.discard_ratio.load(Ordering::Relaxed);
+        let mut rng = rand::thread_rng();
+        rng.gen_ratio(ratio as u32, RATIO_SCALE_FACTOR as u32)
+    }
+
+    pub fn consume(&self, bytes: usize) -> Consume {
+        self.limiter.consume(bytes)
+    }
+
+    pub fn enable(&self, enable: bool) {
+        self.enabled.store(enable, Ordering::Relaxed);
+        if enable {
+            self.tx.send(Msg::Enable).unwrap();
+        } else {
+            self.tx.send(Msg::Disable).unwrap();
+        }
+    }
+
+    pub fn enabled(&self) -> bool {
+        self.enabled.load(Ordering::Relaxed)
+    }
+
+    pub fn is_unlimited(&self) -> bool {
+        self.limiter.speed_limit() == INFINITY
+    }
+}
+
+const SMOOTHER_STALE_RECORD_THRESHOLD: f64 = 300.0; // 5min
+
+// Smoother is a sliding window used to provide steadier flow statistics.
+struct Smoother<const CAP: usize> {
+    records: VecDeque<(u64, Instant)>,
+    total: u64,
+}
+
+impl<const CAP: usize> Default for Smoother<CAP> {
+    fn default() -> Self {
+        Self {
+            records: VecDeque::with_capacity(CAP),
+            total: 0,
+        }
+    }
+}
+
+impl<const CAP: usize> Smoother<CAP> {
+    pub fn observe(&mut self, record: u64) {
+        if self.records.len() == CAP {
+            let v = self.records.pop_front().unwrap().0;
+            self.total -= v;
+        }
+
+        self.total += record;
+
+        self.records.push_back((record, Instant::now_coarse()));
+        self.remove_stale_records();
+    }
+
+    fn remove_stale_records(&mut self) {
+        // make sure there are two records left at least
+        while self.records.len() > 2 {
+            if self.records.front().unwrap().1.saturating_elapsed_secs()
+                > SMOOTHER_STALE_RECORD_THRESHOLD
+            {
+                let v = self.records.pop_front().unwrap().0;
+                self.total -= v;
+            } else {
+                break;
+            }
+        }
+    }
+
+    pub fn get_recent(&self) -> u64 {
+        if self.records.is_empty() {
+            return 0;
+        }
+        self.records.back().unwrap().0
+    }
+
+    pub fn get_avg(&self) -> f64 {
+        if self.records.is_empty() {
+            return 0.0;
+        }
+        self.total as f64 / self.records.len() as f64
+    }
+
+    pub fn get_max(&self) -> u64 {
+        if self.records.is_empty() {
+            return 0;
+        }
+        self.records.iter().max_by_key(|(k, _)| k).unwrap().0
+    }
+
+    pub fn get_percentile_90(&mut self) -> u64 {
+        if self.records.is_empty() {
+            return 0;
+        }
+        let mut v = self.records.make_contiguous().to_vec();
```

Comment:
`make_contiguous` is heavier and unnecessary. I think `.iter().collect()` should work well.

Reply:
Or you needn't call `to_vec`. Just sort the `&mut [T]` returned by `make_contiguous`.